nixos/modules/services/hardware/kanata.nix

```diff
@@ -0,0 +1,173 @@
+            lsft z    x    c    v    b    n    m    ,    .    /    rsft
+            lctl lmet lalt           spc            ralt rmet rctl)
+
+          (defalias
+            ;; tap within 100ms for capslk, hold more than 100ms for lctl
+            cap (tap-hold 100 100 caps lctl))
+        '';
+        description = ''
+          Configuration.
+          See <link xlink:href="https://github.com/jtroo/kanata"/> for more information.
+        '';
+      };
+    };
+  };
+
+  mkName = keyboard: "kanata-${keyboard.name}";
+
+  mkConfig = keyboard:
+    pkgs.writeText "${mkName keyboard}-config.kdb" ''
+      (defcfg
+        ${lib.strings.optionalString keyboard.enableCmd "danger-enable-cmd yes"}
+        linux-dev ${keyboard.device})
+
+      ${keyboard.config}
+    '';
+
+  mkService = keyboard: {
+    name = mkName keyboard;
+    value = {
+      description = "kanata for ${keyboard.device}";
+
+      # Because path units are used to activate service units, which
+      # will start the old stopped services during "nixos-rebuild
+      # switch", stopIfChanged here is a workaround to make sure new
+      # services are running after "nixos-rebuild switch".
+      stopIfChanged = false;
+
+      serviceConfig = hardeningOptions // {
+        DynamicUser = true;
+        SupplementaryGroups = with config.users.groups; [
+          input.name
+          uinput.name
+        ];
+      };
+      script = ''
+        ${cfg.package}/bin/kanata \
+          --cfg ${mkConfig keyboard}
+      '';
+    };
+  };
+
+  mkPath = keyboard: {
+    name = mkName keyboard;
+    value = {
+      description = "kanata trigger for ${keyboard.device}";
+      wantedBy = [ "multi-user.target" ];
+      pathConfig = {
+        PathExists = keyboard.device;
+      };
+    };
+  };
+
+  hardeningOptions = {
+    DeviceAllow = [
+      "/dev/uinput w"
+      "char-input r"
+    ];
+    CapabilityBoundingSet = "";
+    DevicePolicy = "closed";
+    IPAddressDeny = "any";
+    LockPersonality = true;
+    MemoryDenyWriteExecute = true;
+    PrivateNetwork = true;
+    PrivateUsers = true;
+    ProcSubset = "pid";
+    ProtectClock = true;
+    ProtectControlGroups = true;
+    ProtectHome = true;
+    ProtectHostname = true;
+    ProtectKernelLogs = true;
+    ProtectKernelModules = true;
+    ProtectKernelTunables = true;
+    ProtectProc = "invisible";
+    RestrictAddressFamilies = "none";
+    RestrictNamespaces = true;
+    RestrictRealtime = true;
+    SystemCallArchitectures = "native";
+    SystemCallFilter = [
+      "@system-service"
+      "~@privileged"
+      "~@resources"
+    ];
+    UMask = "0077";
+  };
+in
+{
+  options.services.kanata = {
+    enable = mkEnableOption "kanata";
+    package = mkOption {
+      type = types.package;
+      default =
+        if any (keyboard: keyboard.enableCmd) (attrValues cfg.keyboards)
+        then pkgs.kanata-with-cmd
+        else pkgs.kanata;
```

Comment:
considering that it's called `danger-enable-cmd` in the config, would an example suffice here? no idea how often commands are used, but if you have to set an option to change the package anyway it may as well be the package option itself.

Reply:
I am not sure I fully understand your meaning, could you elaborate it?

Reply:
was thinking about removing `enableCmd` on the keyboard definitions, setting the example here to `pkgs.kanata-with-cmd`, and making a note in the description that commands will have to change the package to `kanata-with-cmd` to work. that would keep the module simpler and acknowledge more explicitly that there's multiple packages with different feature sets.

Reply:
In general I agree.

One question: if we remove `enableCmd`, it will be hard to generate the config file.

``` lisp
(defcfg
      ${lib.strings.optionalString keyboard.enableCmd "danger-enable-cmd yes"}
      linux-dev ${keyboard.device})
```

Reply:
yeah, if that's the only option available here it makes sense to keep `enableCmd`. if there are more things one may want to pass to `defcfg` a different option (`extraConfig`?) might be better.

not sure. either way is fine, but keeping the module complexity low does have its benefits.